packages/neo-one-client/src/user/LocalUserAccountProvider.ts

```diff
@@ -823,6 +841,172 @@ export class LocalUserAccountProvider<TKeyStore extends KeyStore, TProvider exte
+          looseAssets = {
+            ...looseAssets,
+            [unspent.asset]: looseAssets[unspent.asset].concat([unspent]),
+          };
+        } else {
+          looseAssets = {
+            ...looseAssets,
+            [unspent.asset]: [unspent],
+          };
+        }
+      });
+
+    const returnOutputs = transactionUnsignedIn.outputs.filter(
+      (output) => common.uInt160ToString(output.address) === addressToScriptHash(from.address),
+    );
+    let newUnspents: ReadonlyArray<UnspentOutput>;
+
+    if (returnOutputs.length > 0) {
+      returnOutputs
+        .filter(
+          (output) =>
+            (looseAssets[common.uInt256ToString(output.asset)] as ReadonlyArray<UnspentOutput> | undefined) !==
+            undefined,
+        )
+        .forEach((output) => {
+          const currentAsset = common.uInt256ToString(output.asset);
+          const toAdd = Math.min(
+            looseAssets[currentAsset].length,
+            Math.floor((byteLimit - getMessageSize()) / newInputSize),
+          );
+          if (toAdd === 0) {
+            return createConsolidatedTransaction();
+          }
+          newUnspents = looseAssets[currentAsset]
+            .sort((coinA, coinB) => coinA.value.comparedTo(coinB.value))
+            .slice(0, toAdd);
+
+          looseAssets = {
+            ...looseAssets,
+            [currentAsset]: looseAssets[currentAsset].slice(toAdd),
+          };
+
+          updatedOutputs = updatedOutputs.map((out) => {
+            if (out.address === output.address && out.asset === output.asset) {
+              return updateOutput(out, newUnspents.reduce((left, right) => left.plus(right.value), new BigNumber('0')));
+            }
+
+            return out;
+          });
+          newInputs = newInputs.concat(this.convertInputs(newUnspents));
+
+          return;
+        });
+    }
+
+    Object.keys(looseAssets)
+      .filter((key) => looseAssets[key].length >= 2)
+      .forEach((key) => {
+        if (byteLimit - getMessageSize() - newOutputSize - newInputSize * 2 > 0) {
+          newUnspents = looseAssets[key]
+            .sort((coinA, coinB) => coinA.value.comparedTo(coinB.value))
+            .slice(
+              0,
+              Math.min(
+                looseAssets[key].length,
+                Math.floor((byteLimit - getMessageSize() - newOutputSize) / newInputSize),
+              ),
+            );
+          const newOutput: Output = {
+            asset: key,
+            value: newUnspents.reduce((left, right) => left.plus(right.value), new BigNumber('0')),
+            address: from.address,
+          };
+          newOutputs = newOutputs.concat(this.convertOutputs([newOutput]));
+          newInputs = newInputs.concat(this.convertInputs(newUnspents));
+
+          return;
+        }
+
+        return createConsolidatedTransaction();
+      });
+
+    return createConsolidatedTransaction();
```

Comment:
This does something. But why a separate function? Just move that code down here.